### PR TITLE
ksud.c:Fix the issue of incompatible __maybe_unused in the GCC compiler kernel used in versions 4.4. x to 4.9. x.

### PR DESCRIPTION
**Pull request by Coconutat**
Not all use GCC compiler kernel support __maybe_unused.Use __maybe_unused will cost compile error.  
Tested different versions of the kernel from versions 4.4 to 4.9.  
Test list:
4.4.x:Xiaomi 6 stock kernel and Huawei hi3660 Proto kernel
4.9.x:OnePlus 6/6T kernel(GCC vsrsion can't use  __maybe_unused,Clang Version can use  __maybe_unused),Huawei Pangu hi3660 Kernel and Xiaomi 8(Clang Version can use __maybe_unused) .

### CODE REVIEW

**Comment by tiann** [comment]
Try __attribute__((unused)) instead

**Comment by utkustnr** [comment]
I was just about to create my own pr for the same issue then I saw this. Beat me by 10 hours nonetheless, anyway

> Try `__attribute__((unused))` instead

__attribute__((unused)) wasn't enough to fool my compiler. I'm using gcc-linaro 7.5-2019.12 toolchain to compile Huawei P20 lite (hi6250) custom kernel 4.9.337. Straight up removing __maybe_unused was enough to complete compilation.

**Comment by Coconutat** [comment]
> Try `__attribute__((unused))` instead

Won't work.Still same issuse.
`
2023-06-19T03:11:36.4161254Z   CC      drivers/kernelsu/ksud.o
2023-06-19T03:11:36.5612864Z ../drivers/kernelsu/ksud.c:120:1: error: expected ‘,’ or ‘;’ before ‘{’ token
2023-06-19T03:11:36.5613279Z  {
2023-06-19T03:11:36.5613457Z  ^
2023-06-19T03:11:36.5620068Z ../drivers/kernelsu/ksud.c: In function ‘ksu_handle_vfs_read’:
2023-06-19T03:11:36.5620646Z ../drivers/kernelsu/ksud.c:336:2: warning: format ‘%d’ expects argument of type ‘int’, but argument 4 has type ‘size_t’ [-Wformat=]
2023-06-19T03:11:36.5621088Z   pr_info("vfs_read: %s, comm: %s, count: %d, rc_count: %d\n", dpath,
2023-06-19T03:11:36.5621332Z   ^
2023-06-19T03:11:36.5621976Z ../drivers/kernelsu/ksud.c:336:2: warning: format ‘%d’ expects argument of type ‘int’, but argument 5 has type ‘size_t’ [-Wformat=]
2023-06-19T03:11:36.5623003Z ../drivers/kernelsu/ksud.c:340:3: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘size_t’ [-Wformat=]
2023-06-19T03:11:36.5623727Z    pr_err("count: %d < rc_count: %d", count, rc_count);
2023-06-19T03:11:36.5623960Z    ^
2023-06-19T03:11:36.5625979Z ../drivers/kernelsu/ksud.c:340:3: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘size_t’ [-Wformat=]
2023-06-19T03:11:36.5629408Z ../drivers/kernelsu/ksud.c:346:3: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘size_t’ [-Wformat=]
2023-06-19T03:11:36.5630075Z    pr_err("copy ksud.rc failed: %d\n", ret);
2023-06-19T03:11:36.5630439Z    ^
2023-06-19T03:11:36.5647496Z ../drivers/kernelsu/ksud.c: At top level:
2023-06-19T03:11:36.5650422Z ../drivers/kernelsu/ksud.c:119:12: warning: ‘count’ used but never defined
2023-06-19T03:11:36.5651000Z  static int count(struct user_arg_ptr argv, int max) __attribute__((unused))
2023-06-19T03:11:36.5651465Z             ^
2023-06-19T03:11:36.5689779Z cc1: warning: unrecognized command line option "-Wno-gcc-compat"
2023-06-19T03:11:36.5692601Z cc1: warning: unrecognized command line option  @"-Wno-int-conversion"
`

**Comment by Coconutat** [comment]
> Try `__attribute__((unused))` instead

Do you have any better idea?

**Comment by tiann** [comment]
The logs shows that the error should not be related to `__maybe_unused`, because there are several other places in the kernel code where `__maybe_unused` is used, so why isn't it giving an error in other places?

**Comment by tiann** [comment]
Additionally, the `__attribute__((unused))` attribute has been supported by gcc since very old versions (probably 2.x), so this should not be an issue.

**Comment by Coconutat** [comment]
> 

Yes,but in KernelSU it won't work.Actually, I think the problem is that if this __attribute__((unused)) or __maybe_unused  is added, the compiler seems unable to handle the content of the {} section.
log：error: expected ‘,’ or ‘;’ before ‘{’ token 
If add __attribute__((unused)) or __maybe_unused,gcc will show it.

**Comment by utkustnr** [comment]
> The logs shows that the error should not be related to `__maybe_unused`, because there are several other places in the kernel code where `__maybe_unused` is used, so why isn't it giving an error in other places?

Believe it or not but by removing it (undoing https://github.com/tiann/KernelSU/commit/c697398893612a882df8f22ac1fb63bae2e79d79) I finished my build. I checked for other occurances as well, you have a point there but only this one gives error on my side for some archaic reason. Afaik this attribute is for compiler to know function might not be called so that it won't cry in the build process. Removing it makes chromeos users mad but old kernel builders happy, keeping it makes the opposite so idk what to purpose.

```
/home/utku/128/android_kernel_huawei_hi6250/drivers/kernelsu/ksud.c:111:1: error: attributes should be specified before the declarator in a function definition
 static int count(struct user_arg_ptr argv, int max) __maybe_unused
 ^~~~~~
cc1: warning: unrecognized command line option ‘-Wno-gcc-compat’
make[3]: *** [/home/utku/128/android_kernel_huawei_hi6250/scripts/Makefile.build:362: drivers/kernelsu/ksud.o] Error 1
make[2]: *** [/home/utku/128/android_kernel_huawei_hi6250/scripts/Makefile.build:693: drivers/kernelsu] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [/home/utku/128/android_kernel_huawei_hi6250/Makefile:1269: drivers] Error 2
make[1]: Leaving directory '/home/utku/128/out'
make: *** [Makefile:153: sub-make] Error 2
```

So to sum it up.
If @Coconutat removes it, he gets another error unrelated to first problem. 
If I remove it, my build completes. 
If it stays, we both get an error but chromeos builders are happy #637.

Quite the mexican standoff we got here.

**Comment by Coconutat** [comment]
> > > 错误：属性应该在函数定义中的声明符之前指定
> 
> 改变这个
> 
> ```
> */
> static int count（struct user_arg_ptr argv，int max）__maybe_unused
> {
> ```
> 
> 到这个
> 
> ```
> */
> __maybe_unused static int count（struct user_arg_ptr argv，int max） 
> {
> ```
> 
> 也适用于我，但idk它如何可能影响其他编译器。[语法是正确的tho](https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Attribute-Syntax)@Coconutat检查这是否适合你。如果没有，我会离开你的头发，并打开另一个公关只是为了这个变化。

Yesterday I had read it,but not to try.(I need to work).

**Comment by Coconutat** [comment]
> > > error: attributes should be specified before the declarator in a function definition
> 
> Welp, changing this
> 
> ```
> */
> static int count(struct user_arg_ptr argv, int max) __maybe_unused
> {
> ```
> 
> to this
> 
> ```
> */
> __maybe_unused static int count(struct user_arg_ptr argv, int max) 
> {
> ```
> 
> also works for me but idk how it might affect other compilers. [Syntax is correct tho](https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Attribute-Syntax) @Coconutat check if this works for you. If not I'll get out of your hair and open another pr just for this change.

Yesterday I had read it,but not to try.(I need to work).

**Comment by Coconutat** [comment]
> > > error: attributes should be specified before the declarator in a function definition
> 
> Welp, changing this
> 
> ```
> */
> static int count(struct user_arg_ptr argv, int max) __maybe_unused
> {
> ```
> 
> to this
> 
> ```
> */
> __maybe_unused static int count(struct user_arg_ptr argv, int max) 
> {
> ```
> 
> also works for me but idk how it might affect other compilers. [Syntax is correct tho](https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Attribute-Syntax) @Coconutat check if this works for you. If not I'll get out of your hair and open another pr just for this change.

Yes,You are right!Brilliant!

File: kernel/ksud.c [hunk 1]
```diff
@@ -108,6 +108,40 @@ static const char __user *get_user_arg_ptr(struct user_arg_ptr argv, int nr)
...
+
+			if (fatal_signal_pending(current))
+				return -ERESTARTNOHAND;
+			cond_resched();
+		}
+	}
+	return i;
+}
+#else
 static int count(struct user_arg_ptr argv, int max) __maybe_unused
```
Comment (tiann) [inline kernel/ksud.c:145]:
```suggestion
__maybe_unused static int count(struct user_arg_ptr argv, int max)
```

Remove all other changes, this should be enough.

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksud.c (hunk 1 above)
Comment (Coconutat) [reply kernel/ksud.c]: Yes,It's ok

**Review by tiann — APPROVED** [review]
